feat(callbacks): introducing callback notifications into the new slack integration

gal-yardeni opened the pull request:
This is the last piece for migrating the old way we used to send slack notifications (via `echo`) to sending those directly via `keel`.
We are using slack SDK to constructing, updating and sending the messages.
In this PR:
- there's a new `SlackCallbackHandler` which is responsible for parsing the response coming back to `keel` (whenever the user clicks on a button inside the message), and then handles the message based on the notification type.
- we are introducing a new confirmation window for the manual judgment notifications
- there's a new endpoint `/notifications/callback/v2` for listening to slack callbacks, which will come directly from `gate`.

Screenshots of the new messages and confirmations:
![Screen Shot 2021-02-09 at 2 54 51 PM](https://user-images.githubusercontent.com/55253849/107560633-6e9b4980-6b92-11eb-9d30-9941ef8a740e.png)
confirmation for promotion:
![Screen Shot 2021-02-09 at 2 55 00 PM](https://user-images.githubusercontent.com/55253849/107560659-78bd4800-6b92-11eb-9b08-979200def4c5.png)
updating the **existing** MJ notifications with the user and action:
![Screen Shot 2021-02-09 at 5 08 13 PM](https://user-images.githubusercontent.com/55253849/107560684-84a90a00-6b92-11eb-8bde-1ff8ae98a59c.png)

gal-yardeni commented on keel-slack/src/test/kotlin/com/netflix/spinnaker/keel/slack/SlackServiceTest.kt:
```diff
@@ -1,80 +0,0 @@
-import io.mockk.verify
-import org.springframework.core.env.Environment
-
-class SlackServiceTest: JUnit5Minutests {
```
there's no value in testing slack services 🤷 . There's also no testing framework for it.

gal-yardeni commented on keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/InteractiveNotificationCallbackController.kt:
```diff
@@ -59,4 +63,22 @@ class InteractiveNotificationCallbackController(
+  fun handleInteractionCallbackV2(request: RequestEntity<String>) {
+    slackCallbackService.handleCallback(request.body)
+
+//    //TODO: check what to do here
```
@luispollo need your input here:
as I'm doing all the logic in the slack module now, how/where should we check the authorization?

luispollo replied:
We can't check Spinnaker authorization here because Slack is totally oblivious to our authorization schema. Instead, this is where we would check the Slack signature. As discussed offline, this is already being done by Wall-e in our deployment at Netflix, so it would be redundant here, but if you want to make it fully OSS-worthy, you could copy that code over from Echo (or look for support in the SDK -- I bet they have it there!).

luispollo replied:
P.S. It looks like you're using the Java Slack Client SDK. They also have another SDK called Bolt, which is even higher level and does the signature validation automagically for you. Here are the setup instructions: https://slack.dev/java-slack-sdk/guides/supported-web-frameworks#spring-boot

luispollo replied:
P.P.S. If you don't wanna switch SDKs (or add the other one), you can add least use the SDK's `Verifier.isValid()` function to verify the signature: https://github.com/slackapi/java-slack-sdk/blob/main/slack-app-backend/src/main/java/com/slack/api/app_backend/SlackSignature.java#L136

gal-yardeni replied:
should I add the verification? I'm not sure how will it plays out with the call from `gate`...

luispollo commented on keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ManualJudgementNotifier.kt:
```diff
@@ -31,32 +32,41 @@ class ManualJudgementNotifier(
   }
 
+  private val isNewSlackEnabled: Boolean
+    get() = springEnv.getProperty("keel.notifications.slack", Boolean::class.java, true)
```
Should we default to `false` if we can't retrieve the config?

gal-yardeni replied:
hmm i'm setting the FP to false in keel, and I plan to remove it pretty soon, not sure if it makes sense to change it now?

luispollo replied:
Up to you... 🙂

luispollo commented on keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ManualJudgementNotifier.kt:
```diff
@@ -31,32 +32,41 @@ class ManualJudgementNotifier(
-        runBlocking {
-          log.debug("Sending notification for manual judgement with config $it")
-          echoService.sendNotification(event.toEchoNotification(it))
+    if (!isNewSlackEnabled) {
```
Maybe a simpler way of doing this would be to make the entire `@Component` conditional on the config property, with a `@ConditionalOnProperty("keel.notifications.slack", havingValue = false, matchIfMissing = true)`.

gal-yardeni replied:
see my comment above - do you think i should change it if i plan to remove it in the next couple of days?

luispollo replied:
Yeah, if you're planning to remove the condition altogether than it's probably not worth it.

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt:
```diff
@@ -240,6 +243,47 @@ class NotificationEventListener(
+        }
+        val currentArtifact = repository.getArtifactVersionByPromotionStatus(config, currentState.environmentName , deliveryArtifact, PromotionStatus.CURRENT)
+
+
```
```suggestion
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/SlackService.kt:
```diff
@@ -68,14 +72,50 @@ class SlackService(
+
+    if (!response.isOk) {
+      log.warn("slack couldn't get username by email. error is: ${response.error}")
+      email
```
```suggestion
      return email
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/SlackService.kt:
```diff
@@ -68,14 +72,50 @@ class SlackService(
+
+    if (!response.isOk) {
+      log.warn("slack couldn't get email by user id. error is: ${response.error}")
+      userId
```
```suggestion
      return userId
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/SlackService.kt:
```diff
@@ -68,14 +72,50 @@ class SlackService(
+    return userId
+  }
+
+  //Update a notification based on the response Url, using blocks (the actual notification).
+  //If something failed, the fallback text will be displayed
```
```suggestion
  // Update a notification based on the response Url, using blocks (the actual notification).
  // If something failed, the fallback text will be displayed.
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,132 @@
+import java.time.Instant
+
+/**
+ * This handler will handel slack callbacks coming from Manual Judgment notifications.
+ * First, it will update the constraint status based on the user response (either approve/reject)
+ * Second, it will construct an updated notification with the action preformed and the user how did it.
```
```suggestion
 * This handler will handle slack callbacks coming from Manual Judgment notifications.
 * First, it will update the constraint status based on the user response (either approve/reject).
 * Second, it will construct an updated notification with the action preformed and the user who did it.
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,132 @@
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  //Updating the constraint status based on the user response (either approve/reject)
```
```suggestion
  // Updating the constraint status based on the user response (either approve/reject)
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,132 @@
+        )
+      )
+
+    //convert status to actual action --> like OVERRIDE_PASS to "approve"
```
```suggestion
    // convert status to actual action --> like OVERRIDE_PASS to "approve"
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,132 @@
+
+    //convert status to actual action --> like OVERRIDE_PASS to "approve"
+    val actionPerformed = actionsMap[actionStatus]
+    if (actionPerformed == null) {
+      log.warn("can't map slack action status to an actual action and therefore can't send an updated notification.")
+      return
+    }
```
Could this ever happen in practice since we convert the `actionStatus` to a `ConstraintStatus` on line 46, which would fail with an exception if it's an unknown value? BTW, you could use `ConstraintStatus` values as the keys to `actionsMap`.

gal-yardeni replied:
I'm not sure if it can happen or not, but when converting the status to the actual actions (like `approve` and so on) with the `actionsMap`, it can be nullable, so I thought to add a check for it.

luispollo replied:
Ok.

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,132 @@
+    respondToCallback(slackCallbackResponse, actionPerformed)
+  }
+
+  fun respondToCallback(slackCallbackResponse: SlackCallbackHandler.SlackCallbackResponse, actionPerformed: String) {
```
nit: import `SlackCallbackHandler.SlackCallbackResponse`.

gal-yardeni replied:
done

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,132 @@
+
+    //construct an update notification
+    val updatedNotification = updateMJNotification(
+      slackCallbackResponse.message["blocks"] as List<Any>,
```
So we're not updating anything in the blocks based on the user's action?

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,132 @@
+      fallbackText)
+  }
+
+  fun updateMJNotification(originalBlocks: List<Any>, fallbackText: String, action: String): List<LayoutBlock> {
```
Consider renaming for clarity:
```suggestion
  fun updateManualJudgementNotification(originalBlocks: List<Any>, fallbackText: String, action: String): List<LayoutBlock> {
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,133 @@
+
+  fun updateMJNotification(originalBlocks: List<Any>, fallbackText: String, action: String): List<LayoutBlock> {
+    try {
+      //This is pretty ugly, but currently slack SDK doesn't have a nice way to parse those fields natively
```
I was looking at some code samples in the Slack documentation and it looks like this could be converted to a `BlockActionPayload` object, perhaps? You could use `ObjectMapper` for that, supposedly.

luispollo replied:
I added another comment down below about built-in support from the SDK for parsing callback requests: https://github.com/spinnaker/keel/pull/1799/files#r574116003

gal-yardeni replied:
done!

gal-yardeni replied:
done!

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/SlackCallbackHandler.kt:
```diff
@@ -0,0 +1,89 @@
+      return
+    }
+
+    val payload = parseSlackPayload(request)
```
From the [Slack Java SDK docs](https://slack.dev/java-slack-sdk/guides/interactive-components), it looks like you don't have to do all this yourself -- it's all built into the SDK.

Example:
```kotlin
app.blockAction("button-action") { req, ctx ->
  val value = req.payload.actions[0].value
  if (req.payload.responseUrl != null) {
    ctx.respond("You've sent ${value} by clicking the button!")
  }
  ctx.ack()
}
```

gal-yardeni replied:
done!

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/SlackCallbackHandler.kt:
```diff
@@ -0,0 +1,89 @@
+  fun handleCallback(request: String?) {
+    if (request == null) {
+      log.error("the request body coming from slack callback is null.")
+      //throw exception?
```
Why are you making the parameter nullable in the first place?

gal-yardeni replied:
do you think it's better to check it in the controller?

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/SlackCallbackHandler.kt:
```diff
@@ -0,0 +1,89 @@
+        // Slack requests use application/x-www-form-urlencoded
+        URLDecoder.decode(body.split("payload=")[1], "UTF-8"))
+    } catch (ex: Exception) {
+      log.warn("error accrued when converting slack callback response to SlackCallbackResponse object.", ex)
```
```suggestion
      log.warn("error converting slack callback response to SlackCallbackResponse object", ex)
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/SlackCallbackHandler.kt:
```diff
@@ -0,0 +1,89 @@
+  }
+
+
+  data class SlackCallbackResponse(
+    val type: String,
+    val user: SlackUser,
+    val actions: List<SlackAction>,
+    val response_url: String,
+    val message: Map<String, Any>
+  )
+
+  data class SlackUser(
+    val username: String,
+    val id: String,
+    val name: String
+  )
+
+  data class SlackAction(
+    val type: String,
+    val action_id: String,
+    val value: String,
+    val action_ts: String
+  )
```
As I mentioned above, it looks like these types are all provided by the SDK and you could replace all this custom parsing with a few lines of code using the SDK.

gal-yardeni replied:
removed!

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/ManualJudgmentNotificationHandler.kt:
```diff
@@ -0,0 +1,98 @@
+  override fun sendMessage(notification: SlackManualJudgmentNotification, channel: String) {
+    log.debug("Sending manual judgment notification for application ${notification.application}")
+
+    with(notification) {
```
```suggestion
    with (notification) {
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/ManualJudgmentNotificationHandler.kt:
```diff
@@ -0,0 +1,98 @@
+                deny("Stop, I've changed my mind!")
+                title("Are you sure?")
+                markdownText("Are you sure you want to promote version ${artifactCandidate.version}?")
+
```
```suggestion
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/ManualJudgmentNotificationHandler.kt:
```diff
@@ -0,0 +1,98 @@
+                markdownText("Are you sure you want to promote version ${artifactCandidate.version}?")
+
+              }
+
```
```suggestion
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/ManualJudgmentNotificationHandler.kt:
```diff
@@ -0,0 +1,98 @@
+              value(ConstraintStatus.OVERRIDE_FAIL.name)
+              actionId("${stateUid.toString()}:${ConstraintStatus.OVERRIDE_FAIL.name}:${type.name}")
+              confirm {
+                markdownText("Are you sure you want to deny version ${artifactCandidate.version}? promotion")
```
```suggestion
                markdownText("Are you sure you want to reject version ${artifactCandidate.version}?")
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/ManualJudgmentNotificationHandler.kt:
```diff
@@ -0,0 +1,98 @@
+              button {
+                text("See changes", emoji = true)
+                url(compareLink)
+                actionId("123")
```
?

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/ManualJudgmentNotificationHandler.kt:
```diff
@@ -0,0 +1,98 @@
+            }
+          }
+        }
+
```
```suggestion
```

luispollo commented on keel-slack/src/test/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandlerTests.kt:
```diff
@@ -0,0 +1,121 @@
+        ),
+        actions = listOf(SlackCallbackHandler.SlackAction(
+          type = "button",
+          action_id = "$constraintUid:OVERRIDE_PASS:MANUAL_JUDGMENT",
```
Nit: this format reads kinda backwards to me. I'd expect `MANUAL_JUDGEMENT:$constraintUid:OVERRIDE_PASS` (so constraint type + contraint uid + status). Also, maybe we could use forward slashes instead of colons to make it more REST-ish? Like `MANUAL_JUDGEMENT/01EQW0XKNR5H2NMPJXP020EQXE/OVERRIDE_PASS`?

gal-yardeni replied:
I'm going to think about the structure a bit more, but I can always refactor this after this PR will get merged.

luispollo commented on keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/InteractiveNotificationCallbackController.kt:
```diff
@@ -59,4 +63,22 @@ class InteractiveNotificationCallbackController(
+   * New endpoint for the new slack integration. This will be called from gate directly instead of echo.
+   */
+  @PostMapping(
+    path = ["/v2"],
```
Consider making this `/notifications/callbacks/slack` for readability.

gal-yardeni replied:
what do you mean? the path is `/notifications/callback/v2` (as I didn't want to override the existing endpoint).

luispollo replied:
I meant that `v2` doesn't really mean anything, and so was suggesting to use `slack` for that last bit. My understanding was that gate calls keel on this endpoint, and so you could make this path whatever you want, right?

gal-yardeni replied:
done!

luispollo commented on keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/InteractiveNotificationCallbackController.kt:
```diff
@@ -59,4 +63,22 @@ class InteractiveNotificationCallbackController(
+    consumes = [MediaType.APPLICATION_FORM_URLENCODED_VALUE],
+    produces = [MediaType.APPLICATION_JSON_VALUE, APPLICATION_YAML_VALUE]
+  )
+  fun handleInteractionCallbackV2(request: RequestEntity<String>) {
```
```suggestion
  fun handleSlackInteractionCallback(request: RequestEntity<String>) {
```

gal-yardeni commented:
@luispollo I've addressed all of your comments :) Let me know if there's anything blocking for merge it tomorrow, I really want to test it in live :)

luispollo commented on keel-slack/keel-slack.gradle.kts:
```diff
@@ -10,6 +10,7 @@ dependencies {
 
   implementation("com.slack.api:slack-api-model-kotlin-extension:1.4.1")
   implementation("com.slack.api:slack-api-client-kotlin-extension:1.4.1")
+  implementation("com.slack.api:bolt-jetty:1.6.0")
```
Hmm, we shouldn't need this because we already have a web server (that's what jetty is).

gal-yardeni replied:
fixed

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/config/SlackConfiguration.kt:
```diff
@@ -1,10 +1,49 @@
-@ConditionalOnProperty("slack.enabled'")
 class SlackConfiguration {
   var token: String? = null
+    //var signature: String? = null
```
```suggestion
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/config/SlackConfiguration.kt:
```diff
@@ -1,10 +1,49 @@
+
+/**
+ * This is a workaround to disable signature verification,
+ * because we are using gate as proxy in front of our server that already checks the Slack signatures.
```
```suggestion
 * because we are using Netflix's Wall-e (Zuul) as proxy in front of Spinnaker and it already checks the Slack signatures.
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt:
```diff
@@ -240,6 +243,46 @@ class NotificationEventListener(
+
+        val deliveryArtifact = config.artifacts.find {
+          it.reference == currentState.artifactReference
+        } ?: return
```
nit: Add a logging statement here if we can't find the artifact.
```suggestion
        }.also { 
          if (it == null) log.debug("Artifact with reference ${currentState.artifactReference}  not found in delivery config" 
        } ?: return
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt:
```diff
@@ -240,6 +243,46 @@ class NotificationEventListener(
+
+        val artifactCandidate = repository.getArtifactVersion(deliveryArtifact, currentState.artifactVersion, null)
+        if (artifactCandidate == null) {
+          log.debug("artifact version is null for application ${config.application}. Can't send failed deployment notification.")
```
```suggestion
          log.debug("$deliveryArtifact version ${currentState.artifactVersion} not found. Can't send manual judgement notification.")
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/SlackService.kt:
```diff
@@ -32,14 +34,12 @@ class SlackService(
                             application: String, type: NotificationType) {
     if (isSlackEnabled) {
-      log.debug("sending slack notification for channel $channel")
+      log.debug("sending slack notification $type for application $application and channel $channel")
```
```suggestion
      log.debug("Sending slack notification $type for application $application in channel $channel")
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,116 @@
+      )
+  }
+
+
```
```suggestion
```

luispollo commented on keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/callbacks/ManualJudgmentCallbackHandler.kt:
```diff
@@ -0,0 +1,116 @@
+  }
+
+
+  fun updateMJNotification(response: BlockActionPayload): List<LayoutBlock> {
```
```suggestion
  fun updateManualJudgementNotification(response: BlockActionPayload): List<LayoutBlock> {
```

luispollo commented on keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/SlackAppController.kt:
```diff
@@ -0,0 +1,35 @@
+import javax.servlet.annotation.WebServlet
+
+@Component
+@WebServlet("/slack/notifications/callbacks")
```
Just as a reminder, we'll need to update gate to match this URI, correct?

gal-yardeni replied:
already updated and deployed it!

luispollo commented on keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/SlackAppController.kt:
```diff
@@ -0,0 +1,35 @@
+) : SlackAppServlet(slackApp){
+  init {
+    //the pattern here should match the action id in the actual button, which is: constraintId:action:Manual_judgment
+    val pattern = "^([a-zA-Z0-9_\\-.]+):([a-zA-Z0-9_\\-.]+):([a-zA-Z0-9_\\-.]+)".toPattern()
```
```suggestion
    val pattern = "^([\w-]+):([\w-]+):([\w-]+)".toPattern()
```

luispollo reviewed (approved):
This is looks so much better, @gal-yardeni! Thanks for your patience working through my suggestions. I left a few comments with mostly nits, but giving you an early approval!
